### ISSUE TITLE
`without_gvl` doesn't seem to work for functions with generic parameters

### ISSUE THREAD

**Issue by oatmealdealer** [Maaarcocr/lucchetto#1]
Hey there - I found this library looking for an answer to the same question posed in https://github.com/matsadler/magnus/issues/91. Essentially, I'm trying to create a Ruby native extension wrapper for a REST API client in Rust that I generated from an OpenAPI spec using [progenitor](https://github.com/oxidecomputer/progenitor). When I realized I was going to need to manually release the GVL from within the Rust extension to allow the API calls to run concurrently, I looked around and didn't find a satisfying answer except for this crate.

To reduce boilerplate in the Ruby native extension, I'm using a wrapper method that takes a `Future` and handles the task of blocking on that future to get a `Result`, then another helper maps that into a `Result<magnus::Value, magnus::Error>` that can be returned to the Ruby side. The only reason this works is that I'm using `serde_magnus` to transcode all responses into Ruby hashes, so any generated API return types (mostly) only need to implement `serde::Serialize`.

So, I think if I could successfully wrap the function that blocks on the `Future` in `without_gvl`, that should allow the Ruby side to make requests concurrently in multiple threads.
 
Here's an example of what I want to do that reproduces the error using this crate and `tokio`:
```rust
use lucchetto::{without_gvl, GvlSafe};
use std::future::Future;

pub struct GvlSafeFuture<'r, O, F>
where
    F: Future<Output = O>,
{
    runtime: &'r tokio::runtime::Runtime,
    future: F,
}

impl<'r, O, F> GvlSafe for GvlSafeFuture<'r, O, F> where F: Future<Output = O> {}

impl<'r, O, F> GvlSafeFuture<'r, O, F>
where
    F: Future<Output = O>,
{
    pub fn run(self) -> O {
        self.runtime.block_on(self.future)
    }
}

#[without_gvl]
fn run_future<'r, O, F>(future: GvlSafeFuture<'r, O, F>) -> O
where
    F: Future<Output = O>,
{
    future.run()
}
```

The `run_future` function will fail to compile with a diagnostic like this:
```
error[E0401]: can't use generic parameters from outer item
  --> src/lib.rs:24:54
   |
23 | #[without_gvl]
   | -------------- help: try introducing a local generic parameter here: `<F>`
24 | fn run_future<'r, O, F>(future: GvlSafeFuture<'r, O, F>) -> O
   |                      -                               ^ use of generic parameter from outer item
   |                      |
   |                      type parameter from outer item
```

I'm wondering if you have any pointers on how I might be able to make this work, since I'm not very familiar with writing Rust macros or `unsafe`. I have a feeling that this may be difficult/a bad idea but it would be awesome if I could figure out a path forward.

**Comment by Maaarcocr** [comment]
Could you provide some additional details about how you plan to pass in a `Future` from ruby? Because in general I would advise to use the macro in this crate around functions that are then exposed to ruby, like in the example: https://github.com/Maaarcocr/lucchetto/blob/b1b69b3ebf5b3414e050c207285c4e55ce24c29b/examples/simple/ext/simple/src/lib.rs#L17

If you can give me more info about this, I can then try to fix it!

**Comment by oatmealdealer** [comment]
Thanks for the response! And sorry, I should have been more specific - I'm not passing in `Future`s from Ruby, it's the calls to the pure-Rust API client that return `Future`s.

The code here is a little messy in its current state, but here's an example of an individual method in my native extension that gets exposed to Ruby:
https://github.com/oatmealdealer/ship_compliant_v1_rb/blob/4c438dc9d3520b2c32fe4792cde026dbb28bf2c5/ext/ship_compliant_v1_rb/src/lib.rs#L135-L143

To give an example of the `Future`s I'm using the generic parameters for, the call from that function to `post_sales_orders_quote_sales_tax_rate` has a return type of
```rust
impl Future<Output = Result<ResponseValue<ShipCompliantRestApiMiddlewaresSwaggerResponseTypesTaxRateSalesOrderByAddressResponse>, Error<ShipCompliantRestApiMiddlewaresSwaggerResponseTypesSalesOrderAddressValidationResponseOrNone>>>
```

That `Future` gets passed to `fn call`, the part of which I want to modify is:
https://github.com/oatmealdealer/ship_compliant_v1_rb/blob/4c438dc9d3520b2c32fe4792cde026dbb28bf2c5/ext/ship_compliant_v1_rb/src/lib.rs#L101

`self.runtime.block_on(f)` is the function call around which I want to release the GVL. Extending my original example, I'm thinking I would add something like this to `impl V1Client`:
```rust
    fn run_future_without_gvl<F, T, E>(&self, f: F) -> F::Output
    where
        T: Serialize,
        E: Serialize + Debug + Send + Sync,
        ship_compliant_v1_rs::Error<E>: Display + Debug,
        F: std::future::Future<Output = Result<ResponseValue<T>, ship_compliant_v1_rs::Error<E>>>,
    {
        run_future(GvlSafeFuture { runtime: &self.runtime, future: f })
    }
```

I might still not be explaining this super well, but let me know if that makes sense.

**Comment by Maaarcocr** [comment]
I would suggest to use create a new function, called something like `call_without_gvl` which does not return a Future, but instead it returns the result of the future in a sync way and annotate that with `without_gvl`. Keep in mind that you should not use any ruby in it and it should essentially only block on the future you want to run. Then in `call` you can transform the result of it into a ruby value.

**Comment by oatmealdealer** [comment]
> I would suggest to use create a new function, called something like `call_without_gvl` which does not return a Future, but instead it returns the result of the future in a sync way and annotate that with `without_gvl`. Keep in mind that you should not use any ruby in it and it should essentially only block on the future you want to run. Then in `call` you can transform the result of it into a ruby value.

Right, that's what I'm trying to do with the `run_future` function in my original post. The problem is that the `run_future` function needs to have generic type parameters, but then I get the error `can't use generic parameters from outer item`. I looked at the code of `lucchetto` a little and I think maybe it's related to the anonymous wrappers around the function, but I'm not sure.

If you want to take a look locally, here's my example from earlier as a library crate: https://github.com/oatmealdealer/lucchetto-test
